### DISCUSSION TITLE
🐞 Codegen changing causing inconsistent func names across Go SDK module and dependency in case of protocol in name

### DISCUSSION THREAD

**Discussion by jpadams** [dagger/dagger#12329] [maintainers]
### What is the issue?
tl;dr I suspect that codegen is trying to make certain keywords or acronyms consistent and thus using `TCP` instead of `Tcp`, for example. We need to look into that, ensure we know and like the behavior, and document it. My suspicion seems to be confirmed when I use Foo instead of protocol names (see below). This has the effect that a Go SDK function dependency might have a `TcpTest` function, but to call it in the module that installed the dependency, you'd have to use `TCPTest` which is very confusing. We document these cases and perhaps should alert or fail on the use of names that will cause this issue (or alternatively change the behavior in codegen).

I suspect this is due to the "Common Initialisms" used by`gqlgen`
https://github.com/dagger/dagger/blob/main/go.mod#L9
https://github.com/99designs/gqlgen/blob/master/codegen/templates/templates.go#L709-L763
@helderco please confirm
@vikram-dagger if this is the case, we could document the initialisms to watch out for. Not sure if there are similar initialisms in Python or Typescript or other SDKs.

When you use a function name that contains a protocol name like TCP, UDP, IP, HTTP, HTTPS, many Golang users would assume they should use PascalCase for their public function names, for example, `TcpTest` or `HttpsTest`, but when those function names are run through codegen in the Go SDK at least, certain substrings (containing protocol names, for example) are upcased. So the functions above become `TCPTest` and `HTTPSTest`. This has the effect that a Go SDK function dependency might have a `TcpTest` function, but to call it in the module that installed the dependency, you'd have to use `TCPTest` which is very confusing.

On the other hand, a function name like `FooTest` results in `FooTest` in the codegen.

### Dagger version

dagger v0.11.0 mac

### Steps to reproduce

https://discord.com/channels/707636530424053791/1229172690082857173/1229172690082857173

https://github.com/jpadams/main/blob/main/dep/main.go
```golang
package main

type Dep struct{}

func (m *Dep) TcpTest() string {
    return "TcpTest"
}
func (m *Dep) UdpTest() string {
    return "UdpTest"
}
func (m *Dep) IpTest() string {
    return "IpTest"
}
func (m *Dep) HttpTest() string {
    return "HttpTest"
}
func (m *Dep) HttpsTest() string {
    return "HttpsTest"
}
func (m *Dep) FooTest() string {
    return "HttpsTest"
}
```
```shell
git clone https://github.com/jpadams/main
cd main
```
```shell
# Tcp -> TCP
grep -rin tcptest | grep func
./internal/dagger/dagger.gen.go:1967:func (r *Dep) TCPTest(ctx context.Context) (string, error) {
./dep/main.go:5:func (m *Dep) TcpTest() string {

# Https -> HTTPS
grep -rin httpstest | grep func
./internal/dagger/dagger.gen.go:1894:func (r *Dep) HTTPSTest(ctx context.Context) (string, error) {
./dep/main.go:17:func (m *Dep) HttpsTest() string {

# versus FooTest
grep -rin footest | grep func
./internal/dagger/dagger.gen.go:1870:func (r *Dep) FooTest(ctx context.Context) (string, error) {
./dep/main.go:20:func (m *Dep) FooTest() string {
```
Or open `main.go` in VSCode
<img width="300" src="https://github.com/dagger/dagger/assets/3187222/c3fd377e-f91c-4746-bf95-feda8993a690" />

**Comment by helderco** [comment]
> I suspect this is due to the "Common Initialisms" used by`gqlgen`

You're right that it originally came from there, but we've managed our own fork since the very beginning of the Go SDK, in codegen. There's a couple Dagger additions as well: 

https://github.com/dagger/dagger/blob/bdce95d0abfa9014a35f5e3962a64afce03d5fca/cmd/codegen/generator/go/templates/lint_name.go#L122-L124

This is the underlying problem of:
- https://github.com/dagger/dagger/issues/6669

Difference is you're talking about the choice of function names, while that's about the struct's name (module name). So a module initialized with `--name=ssh` created a struct named `Ssh`. In that particular case, there was a compilation failure which was fixed in https://github.com/dagger/dagger/pull/6692, but you make a good point:

> This has the effect that a Go SDK function dependency might have a `TcpTest` function, but to call it in the module that installed the dependency, you'd have to use `TCPTest` which is very confusing.

I agree. With the introduction of modules and putting API schema expansion on the hands of users, doing the initialisms doesn't seem worth the potential confusion. At least not in an automatic way like today.

Here’s another fun one for you:
- https://github.com/dagger/dagger/issues/6310

The way I’ve wanted to fix the inconsistency is through:
- https://github.com/dagger/dagger/issues/5226

You can see [`nameWords()` in the codegen module](https://daggerverse.dev/mod/github.com/helderco/daggerverse/codegen@357dfd7f97eae9a62a2a13454b48cb7c69c44e34#Field.nameWords):
```console
❯ dagger -m github.com/helderco/daggerverse/codegen@v0.1.3 call introspect get-type --name=Container fields name-words
as service
as tarball
build
default args
directory
entrypoint
env variable
env variables
experimental with all GPUs
experimental with GPU
export
exposed ports
file
from
id
image ref
import
label
labels
mounts
pipeline
platform
publish
rootfs
stderr
stdout
sync
terminal
user
with default args
with default terminal cmd
with directory
with entrypoint
with env variable
with exec
with exposed port
with file
with files
with focus
with label
with mounted cache
with mounted directory
with mounted file
with mounted secret
with mounted temp
with new file
with registry auth
with rootfs
with secret variable
with service binding
with unix socket
with user
with workdir
without default args
without entrypoint
without env variable
without exposed port
without focus
without label
without mount
without registry auth
without unix socket
without user
without workdir
workdir
```

The intent there is for SDKs to codegen based on this common base, rather than the lower level GraphQL introspection and everyone figuring out and implementing the same rules all over. 

Notice how `GPU` capitalization is preserved:
```
experimental with all GPUs
experimental with GPU
```

This way it’s easier to convert to any naming convention because you don’t have to worry about preserving capitalization in any non-first letter:
- `camelCase`: capitalize only the first letter of second to last words, and concatenate them
- `PascalCase`: capitalize only the first letter of every word, and concatenate them
- `snake_case`: lower case every word and join with underscore
- `kebab-case`: lower case every word and join with dash

Example (not how it looks today!):
- GraphQL: `experimentalWithAllGPUs`
- Go: `ExperimentalWithAllGPUs`
- Python: `experimental_with_all_gpus` (today it ends in `gpu_s`!)
- TypeScript: *same as GraphQL*

But, codegen here works off of the API, so we need to try and persist the capitalization in acronyms in user modules when creating the API fields, like the above example.

So, let’s say the core API defines `ContainerID`, everything works the same from codegen, but Go would preserve the capitalization rather than force it. If a module creates a `GetContainerID`, it would produce `getContainerID` in the API. But if the module uses `GetContainerId`, that would create a `getContainerId` in the API. In both cases it would look the same in the module and in a generated client.

The problem now is it’s a breaking change for Go and Typescript modules that depend on a module that uses acronyms without being capitalized (like `TcpTest`).

**Comment by TomChv** [comment]
I removed the TS label since it's not related to TypeScript

**Comment by nipuna-perera** [comment]
This came up in my use case too. Here's the relevant Discord chat - https://discord.com/channels/707636530424053791/1120503349599543376/1357053921931296911

> I didn't realize this was the case before but I got this feedback from a code review. I have a function named `WithOsUtilities`. That translates to
`with-os-utilities` in the CLI and `WithOsUtilities` in the SDK. However, the Go naming convention is `WithOSUtilities` and linters will throw a warning on it. 
\
So if I change my function to `WithOSUtilities`, the resulting CLI is `with-osutilities` and SDK is `WithOsutilities`. CLI is awkward but manageable but the SDK generated function still doesn't adhere to Go convention. 
\
Is there a trick I'm missing? Or is it a design decision/limitation?

The change @helderco suggested makes sense to me and I'd much rather have a breaking change now than later when I've written a lot more Dagger :) Any chance this can be prioritized?
